perf/config.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Configuration for perf event manager.
+package perf
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+
+	"k8s.io/klog"
+)
+
+type Events struct {
+	// List of perf events to be measured. Events in this list
+	// will not be grouped. See group_fd argument documentation
+	// at man perf_event_open.
+	NonGrouped []Event `json:"non_grouped"`
+
+	// List of groups of events to be measured. See group_fd
+	// argument documentation at man perf_event_open.
+	Grouped [][]Event `json:"grouped"`
```

Comment:
Can you help me understand of the value of having groups:

1. Each event in the group still gets its own metric stream, right?
2. Grouping event A and B just impacts scheduling to _somehow_ make metrics better.  From the man page: 

`An event group is scheduled onto the CPU as a unit: it will be put onto the CPU only if all of the events in the group can be put onto the CPU.  This means that the values of the member events can be meaningfully compared—added, divided (to get ratios), and so on—with each other, since they have counted events for the same set of executed instructions." `

I don't quite understand how Events can be scheduled by the kernel.  I thought these "Events" were things we were looking for the occurrence of, not things we were running (i.e. scheduling)

Reply:
Event grouping is extremely useful when you want to calculate a derivative metric - it is guaranteed that events that you collect were calculated in exactly same period of time.
Kernel is responsible for "scheduling" - it instructs hardware performance counters (that are just pieces of silicon) to react to occurrence of specific perf events.

Reply:
I'll try and rephrase to see if I get it now:  The kernel is essentially making the process of _updating_ the group of performance counters atomic so you can't ever measure a time where one counter is updated, but others are not?

Does it also change the metric you get back?  I.E. is there a reason why we don't just have one large group?

Reply:
Yes, I think I that atomicity is very good name for it :) 
When you want to assure atomicity then you can't measure more events that you have hardware perf counters available (8 per physical core/4 per HT on current generation Xeons) so using single big group is simply impossible. If you are antes to measure more events that you have counters available then your events (or group of them) will be multiplexed and results will have to be scaled.

Reply:
Really appreciate the explanation.  It all makes sense now.

If I were to create a group of 100 perf counters, how would cAdvisor fail?  Is there a way we can fail on startup if it won't work?

Reply:
In such case syscall (perf_event_open) will fail. I suppose that EINVAL will be returned so the error itself will not be very descriptive. 
It is possible to determine properties of hardware counters from cpuinfo, though and such information could be used to validate configuration before calling perf_event_open.

Reply:
Or even just calling perf_event_open with the config on the root cgroup during startup would be good.  That way we can crash cAdvisor if it isn't valid